Fix ignored signature constraints in approximation of recursive modules signatures

clementblaudeau opened the pull request:
# Overview 
Both module constraints and type constraints are ignored during the approximation phase (when typing recursive modules), resulting in wrong approximated signatures:

```ocaml
module type Test = sig
  module Empty : sig end

  module rec X : sig
           module A : sig type t end
           include (sig module A : sig end end) (* could shadow A *)
                   with module A := Empty (* actually, no shadowing, as A is removed (destructive substitution) *)
           type u = A.t
         end
     and Y : sig type u = X.A.t end
end
```
fails with
```ocaml
Line 11, characters 26-31:
11 |      and Y : sig type u = X.A.t end
                               ^^^^^
Error: Unbound type constructor X.A.t
```
This comes from the fact that the destructive constraint on the included signature is not properly computed in the approximation phase, leaving the `module A : sig end` field inside, which in turns shadows the original `module A : sig type t end`, making `X.A.t` undefined.  

# Context
This PR fixes the same problem as [this PR in Oxcaml](https://github.com/oxcaml/oxcaml/pull/4121). While the fix proposed there seems fine, it does not benefit from the refactoring of #13911 (as OxCaml was based on 5.2). The added test-file was borrowed from there. The case of module-type constraints was already fixed by #13829.

# Fix 

Merging module constraints in approximation mode now uses an `~approx` flag, similarly to module type constraints: it disable equivalence and wellformedness checks, but performs the merging. For type constraints, a special `merge_type_approx` function with a simplified logic is defined. It still uses the normal merging infrastructure. 
Overall, a better skeleton for the approximated signature is build, as the destructive substitutions are correctly removing the fields. 

# Notes 

* This PR changes the `lookup_module_path ~load:false` for `lookup_module`, where the load flag is not set to false. It should not have much impact for well-typed programs, as the loading would happen after the approximation phase anyway.
* Ill formed substitutions where the field is not present are caught at approximation phase, while other forms of ill-formedness are caught at typechecking (when the equivalence and wellformedness checks are not disabled)

# ~~Missing:~~ 
- [x] Add similar support for type constraints
- [x] Add more tests
- [x] Update changelog entry with reviewer

clementblaudeau committed:
Add the approx flag to merging of module constraints

- While module types constraints were using the [approx] flag to disable
equivalence checking when merging, module constraints were only checked for
cyclicity. Now, both use the same logic: the constraints is approximated and
then merged in approx mode, where no equivalence check is done. It computes a
better skeleton for the approximated signature, as the destructive substitutions
are correctly removing the fields.

- Add tests borrowed (and adapted) from
https://github.com/oxcaml/oxcaml/pull/4121

- This commit changes the [lookup_module_path ~load:false] for [lookup_module],
where the load flag is not set to false. I should not have much impact for
well-typed programs, as the loading would happen after the approximation phase
anyway.

Octachron commented on typing/typemod.ml:
```diff
@@ -1083,15 +1087,20 @@ and approx_constraint env body constr =
-  | Pwith_module (_, lid') ->
+  (* module substitutions are approximated and merged, checking for cyclicity *)
+  | Pwith_module (id, lid)
+  | Pwith_modsubst (id, lid) ->
```
The `Pwith_typesubst` should be handled too:
```ocaml
module Empty = struct end
module rec X: sig
  module type T = sig type t end with type t := int
  module F(A:T): sig type t end
end = struct
  module type T = sig end
  module F(A:T) = struct type t end
end
and Y: sig type t = X.F(Empty).t end = struct
   type t = X.F(Empty).t
end
```

clementblaudeau replied:
Addressed by 81a88d7013d58c765715ef3b8191e7985862b4c9. This specific example works. Should I add it to the testsuite ?

Octachron replied:
I think so because triggering the bug requires an interaction of non common features (application paths, recursive modules and with constraints).

clementblaudeau replied:
Added in 72d597a268

clementblaudeau commented:
I've updated the description of the PR. I've extended the scope of the PR to also deal with the similar issue for type contraints (the last missing form of constraint).

clementblaudeau committed:
Add a [merge_type_approx] for approximation of type constraints

This commit introduces a new function to specifically deal with merging type
constraints in approximation mode:
- destructive constraints actually remove the constrained field, to prevent
incorrect approximation (specifically, incorrect shadowing)
- non-destructive constraints are treated as an identity patch, where the
constrained field is replaced by itself. This allows us to reuse the normal
merging infrastructure and fail early in case of ill-formed constraints where
the field is not present (other forms of ill-formedness are caught later)

The [post_process] function is made aware of the approximation flag to disable
wellformedness checks

clementblaudeau committed:
Update Changelog

clementblaudeau committed:
Add tests for module and type constraints during approximation

clementblaudeau commented:
I've updated (again) the PR description to also include the type constraints. 

One notable choice I made was to **not** ignore *non-destructive* type constraints, but to do an identity patch (replacing the field by itself). This is made to have a more uniform treatment (with regard to module and module type constraints), and catch ill-formed constraints early (when the field does not exist).

clementblaudeau committed:
Add another test for destructive type constraints approximation

Octachron commented on testsuite/tests/typing-modules/merge_constraint.ml:
```diff
@@ -617,3 +617,237 @@ Lines 4-8, characters 18-26:
+
+
+(* Approximating a signature with a module constraint should merge the module
+   (abstract, shallow, non destructive case) *)
```
shallow ⇒ deep

clementblaudeau replied:
Fixed in 5d986309bc333b5c4c9628092d56c61c058ef07c

Octachron commented on testsuite/tests/typing-modules/merge_constraint.ml:
```diff
@@ -617,3 +617,237 @@ Lines 4-8, characters 18-26:
+   caught during approximation.  *)
+module type IllFormed_Type_No_Field = sig
+  module rec X : (sig
+                   type t
```
I would propose to add a class named `v` to test that with is not trying to remove ghost item.

clementblaudeau replied:
Addressed by 5d986309bc333b5c4c9628092d56c61c058ef07c

Octachron commented on typing/typemod.ml:
```diff
@@ -1070,28 +1110,39 @@ and approx_modtype_info env sinfo =
+    | Pwith_typesubst _
+    | Pwith_modtypesubst _
+    | Pwith_modsubst _ -> true
+    | _ -> false
```
This form is brittle, I would propose to make this `match` a separate function,  enumerate all cases in this function and share it between `transl_with` and `approx_constraint`.

clementblaudeau replied:
Addressed by 5d986309bc333b5c4c9628092d56c61c058ef07c. I'm also considering proposing in another PR a refactor that would merge the `Pwith_` constructs two by two and add a destructive flag.

Octachron reviewed (approved):
The fix is correct and make the behavior of approximate signature for recursive modules much more regular with very little increase in term of code complexity.

clementblaudeau committed:
Address review comments on #14100

- Factor out computation of the [destructive] flag for constraints between
merging in normal and approximation modes

- Fix typos in test descriptions

- Add test that ghost items are not affected by constraints

- Update Changelog

clementblaudeau commented:
I've addressed your comments and updated the changelog.

clementblaudeau commented:
I've rebased on trunk to avoid the merge conflict with the testsuite added in #14117

Octachron commented on Changes:
````diff
@@ -78,6 +78,22 @@ Working version
+  module rec X : ((sig module A : sig end end) with module A := X0)
+  and Y : sig type t = X.A.t end (* Unbound type constructor X.A.t *)
+  ```
+
````
It is better to avoid a newline here (at the very least for my tool parsing changelog entries).

clementblaudeau replied:
Removed in 511b6a45bb

Octachron commented on Changes:
```diff
@@ -78,6 +78,22 @@ Working version
+  Now, type and module constraints are properly merged during approximation ,
+  reusing the infrastructure for normal merging of constraints, but disabling
+  any wellformedness check.
+  (Clement Blaudeau, review by Florian Angeletti)
```
I think it might be nice to cite @rtjoa as the inspiration or initial reporter for the bug fix.
(Except of course if you object to the mention @rtjoa ).

clementblaudeau replied:
I've added @rtjoa as an author (initial bug report + inspiration for the solution). @rtjoa, feel free to tell me if you want to be mentioned differently (511b6a4)

rtjoa replied:
That is gracious, appreciated. And thanks for the fix :slightly_smiling_face:

clementblaudeau committed:
Update Changelog
